The menu

mjaskula opened the pull request:
~I don't think that this is ready to merge, but I wanted to get feedback on where I am to see if you all like the direction I'm taking.~ This branch has been updated with early feedback and is stable and ready for review. I've updated the test below to indicate changes made.

The program displays a menu allowing the user to select the program that they'd like. Upon selection, the menu's handlers are cleared and new handlers are added to detect the exit command. Then the selected program is imported. ~The menu thread hangs around waiting for the exit command. Its impact should be minimal as it's using an ``Asyncio.Event`~ Asyncio is no longer used. No thread is created. A normal button handler is used to detect the exit request.


- ~This approach forces script authors to use some ``asyncio`` code in their scripts. I believe that the impact is minimal and I've documented the requirements thoroughly.~ In addition I've provided an example ``hello_world.py``  which can serve as a reference as well as a starting point for new scripts.
- I've updated all of the scripts except consequencer to support the menu. The changes were minimal, despite what the diffs look like. I'm happy to put each of the script's changes in their own PR if you think it make more sense to discuss them individually.
- The 'copy all the code to the pico' problem isn't really solved, but I'm hoping that issue #71 makes it much easier. (it has)
- ~The classes ``LongPressAndTapButton`` and ``Menu`` are written to be abstract enough to be used elsewhere. If we find another place to use them they should be moved to ``europi.py``~ ``LongPressAndTapButton`` has been removed and ``Menu`` has been put into the new ``ui.py`` module in firmware.


Known remaining work:
- ~save state so that the menu automatically loads the last running program~
- ~update the consequencer~
- ~a couple small TODOs~
- ~document that ``reset_state()`` shouldn't be called in scripts~

awonak commented on software/contrib/menu.md:
```diff
@@ -0,0 +1,142 @@
+* button 1: scroll up
+* button 2: scroll down
+* tap button 1 while holding down button 2: launch selected program
+* tap button 2 while holding down button 1: launch selected program
```
This interface felt a little awkward to me. How do you feel about changing it to use a knob to scroll through the list of scripts, then a single button press to select? I imagine if the number of scripts gets large then repeatedly tapping the button to find a program could get tedious.

mjaskula replied:
That sounds like a good interface if we had an encoder, especially if the encoder had a 'press' ability (is that standard?). With the knobs it sounds a little weird in that the initial selection would be on whatever lined up with wherever the knob happened to be positioned when the menu launched. I feel like that would be unexpected for the user. Also if we had _a lot_ of programs it may be hard to select with a knob.

I can give it a try and see.

What I do like about that is that it could easily enable a hierarchy (folders), with one button being 'enter' and the other being 'back'.

awonak commented on software/contrib/menu.md:
```diff
@@ -0,0 +1,142 @@
+* button 1: scroll up
+* button 2: scroll down
+* tap button 1 while holding down button 2: launch selected program
+* tap button 2 while holding down button 1: launch selected program
```
This interface felt a little awkward to me. How do you feel about changing it to use a knob to scroll through the list of scripts, then a single button press to select? I imagine if the number of scripts gets large then repeatedly tapping the button to find a program could get tedious.

awonak commented on software/contrib/menu.md:
```diff
@@ -0,0 +1,142 @@
+## Menu Inclusion
+
+In order to be included in the menu a program needs to meet a few additional requirements, described below. Programs are
+not required to participate in the menu, but if they do not, they should be added to the ``EXCLUDED_SCRIPTS`` list near 
```
Perhaps another possible approach would be to provide a base class that defines expected methods and behavior, then you can pull the list of script classes that implement the base class for the list of menu scripts instead of requiring users to declare exclusion.

https://docs.python.org/3/library/stdtypes.html?highlight=__subclasses__#class.__subclasses__

https://docs.python.org/3/library/stdtypes.html?highlight=__subclasses__#definition.__qualname__

Here's a drastically oversimplified explanation of my idea:
```python

class MenuScript:
  def main(self):
    """helpful documentation goes here"""
    raise NotImplementedError

class HelloWorld(MenuScript):
  def main(self):
    # properly implemented here

## in the menu.py
import MenuScript

# List scripts using their qualified name
for cls in MenuScript.__subclasses__():
    print(cls.__qualname__)
    # if selected, run that class
    asyncio.create_task(self._run_script(cls))

```

mjaskula replied:
Yeah, I thought about that, as mentioned in the original issue, and I do think that it would have a lot of benefits. I went in this direction because it would have the lowest impact on the existing scripts, and more closely matches how we are creating scripts today.

I expect that the use of the exclusion list would be very limited. This would be something that we check for when accepting submitted scripts. Either the script follows the requirements in this doc or it is added to the exclude list.

I can work up a PR with the base class idea if you like, so that we can see what it could look like. My suspicion would be that it's a little more complicated for not much more benefit.

mjaskula replied:
Sounds good to me. My gut says that a base class was the way to go, but the impression that I got from the issue's discussion was that we wanted this to be as few changes as possible.

Anyway, I look forward to seeing your base class implementation.

mjaskula replied:
Base class added

awonak reviewed (commented):
Dropping in some quick thoughts. I'd still like to explore this more and take our time to get it right.

Excellent job so far, no changes to the firmware, but it still puts some requirements on users not participating in the Menu system. I have added some suggestions that may help simplify the way we gather and execute menu scripts.

awonak commented on software/contrib/menu.md:
````diff
@@ -0,0 +1,141 @@
+    assert increment(1) == 2
+```
+
+1. **Import ``uasyncio`` but fall back to ``asyncio``:** Since ``uasyncio`` doesn't exist in CPython, it will raise an 
````
I _think_ this may be unnecessary. I noticed in the MicroPython 1.17 release notes:

> u-module references have been replaced with just the module name without the u-prefix to help clear up the intended usage of modules in MicroPython.

https://github.com/micropython/micropython/releases/tag/v1.17

awonak replied:
Hmm, nope. Just tested this and got:
```
Traceback (most recent call last):
  File "main.py", line 5, in <module>
ImportError: no module named 'asyncio'
```

mjaskula commented on software/contrib/menu.py:
```diff
@@ -0,0 +1,184 @@
+    - b2: choose the selected item
+    """
+
+    def __init__(self, items, select_func):
```
yeah, that's sounds like a good idea. I'll add it.

mjaskula replied:
Done.

mjaskula commented:
Made the following changes:
- remove asyncio
- renamed menu.py to bootloader.py and put it in with the firmware
- moved the Menu class to a new ui.py module in firmware
- removed the LongPressAndTapButton class, added long both button support to europi.py
- added the EuroPiScript base class to firmware and updated all scripts to use it


Open work:
- update all docs with the above changes
- make sure it all works with the new packaging

roryjamesallen commented on software/firmware/ui.py:
```diff
@@ -0,0 +1,50 @@
+
+    def _inverted_text(self, s, x, y):
+        """displays the given text with an inverted background"""
+        oled.fill_rect(x, y, CHAR_WIDTH * len(s), CHAR_HEIGHT, 1)
```
Suggested change so that the highlighted script text has a border of 1 pixel
``oled.fill_rect(x, y-1, CHAR_WIDTH * len(s), CHAR_HEIGHT+2, 1)``

roryjamesallen replied:
Suggested change so that the highlighted script text has a border of 1 pixel
``oled.fill_rect(x, y-1, CHAR_WIDTH * len(s), CHAR_HEIGHT+2, 1)``

roryjamesallen commented on software/contrib/menu.md:
```diff
@@ -0,0 +1,52 @@
+1. **Import ``EuroPiScript``:** Import the ``EuroPiScript`` base class so that we can implement it below.
+2. **Define a class that extends ``EuroPiScript``:** This defines your script as one that adheres to the menu's requirements.
+3. **A ``main()`` method:** The main code of the script should be encapsulated in an 'main()' method. The menu will initialize your class and then call this function to launch your script.
+4. **a ``__name__`` guard around the ``main()`` call site:** This allows your code to work correctly when it itself is the ``main.py``
```
I'd suggest adding either a link to an explanation of name guarding (https://docs.python.org/3/library/__main__.html), or a note that all you need to do is copy and paste ```if __name__ == "__main__":``` into your own program.
I'm just worried about beginner users not understanding the purpose/usage of the line

roryjamesallen commented on software/firmware/ui.py:
```diff
@@ -0,0 +1,50 @@
+        oled.text(s, x, y, 0)
+
+    def draw_menu(self):
+        """This funciton should be called by your script's main loop in order to display and refresh the menu."""
```
Spelling: 'funciton' to 'function'

roryjamesallen commented on software/programming_instructions.md:
```diff
@@ -49,15 +49,16 @@ Use the exact same process as for the ssd1306 library to install the europi libr
 * [Option 1](#write-your-own-program-from-scratch): Start writing your own program from scratch
 * [Option 2](#copy-someone-elses-program-to-run-on-your-module): Use someone else's program from the [contrib folder](/software/contrib/)
-* [Option 3](#calibrate-the-module): Calibrate the module for higher accuracy
+* [OPtion 3](#install-the-contrib-scripts-and-setup-the-menu): Install all of the contrib scripts and use the bootloader menu
```
Spelling: 'OPtion' to 'Option'

roryjamesallen commented on software/programming_instructions.md:
```diff
@@ -91,7 +92,18 @@ Now you have access to the inputs and outputs using easy methods, which you can
+### Install the contrib scripts and setup the menu
+
+1. Use the exact same process as for the ssd1306 and europi libraries to install the europi-contrib library:
+   1. In Thony, click Tools -> Manage Packages to open the package manager.
```
Spelling: 'Thony' to 'Thonny'

roryjamesallen commented on software/programming_instructions.md:
```diff
@@ -91,7 +92,18 @@ Now you have access to the inputs and outputs using easy methods, which you can
+   3. Click the result named 'micropython-europi-contrib'.
+   4. Click 'Install'.
+   5. You will now see the new directory 'contrib' inside the 'lib' folder, containing the contrib scripts.
+2. The menu program is called ``bootloader.py`` and can be found in the lib directory on the pico or in the [``software/firmware``](/software/firmware/bootloader.py) directory on github. Copy this file as your ``main.py`` as described in 
```
I think this is outdated, there now seem to be two separate files, ``bootloader.py`` which should not be altered, and ``menu.py`` which should be.
This segment (I think) should read:
"The menu program is called ``menu.py`` and can be found in the lib directory on the pico or in the [``software/contrib``](/software/contrib/menu.py) directory on GitHub."

awonak commented on software/firmware/bootloader.py:
```diff
@@ -0,0 +1,90 @@
+                    old_selected = self.menu.selected
+                    self.menu.draw_menu()
+                time.sleep(0.1)
+            self._save_state(self.run_request)
```
Another size friendly option for saving more complex data here is python `struct`

https://docs.python.org/3/library/struct.html

Take a look at this example I did a while back:

https://github.com/awonak/EuroPi-Prototype/blob/main/contrib/expander/clock_distributor_europi.py#L81
https://github.com/awonak/EuroPi-Prototype/blob/main/contrib/expander/clock_distributor_expander.py#L60

awonak reviewed (commented):
I'd suggest updating the main description of this PR to match the state things have currently landed at.

I'll test this out later today and let you know how it goes!

mjaskula commented on software/firmware/bootloader.py:
```diff
@@ -0,0 +1,90 @@
+    """
+
+    def __init__(self, script_classes):
+        EuroPiScript.__init__(self)
```
Ahh, this is a hang-over from when I was using multiple inheritance. Will fix.

awonak commented:
Excellent work here! I'm excited for the improvements this change will bring.
